cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll

```diff
@@ -13,6 +13,11 @@ import semmle.code.cpp.models.interfaces.DataFlow
  */
 private class StdBasicString extends TemplateClass {
   StdBasicString() { this.hasQualifiedName("std", "basic_string") }
+
+  Declaration getAnInstMemberNamed(string name) {
+    result = getAnInstantiation().getAMember() and
+    result.hasName(name)
```

Comment:
Does this work for members that are templates? There's a few template member functions in the string library, but I don't think it has any template classes or template variables that are members of other templates.

Reply:
(discussing elsewhere: this is probably OK for std::string but we may need to adapt the pattern to work elsewhere)

Reply:
The new pattern now looks like this:
```
this.getDeclaringType() instanceof StdBasicString and
this.hasName("push_back")
```
I think if the object we're modelling is itself a template instantiation it would be a template function instantiation, I don't think those have anything (like `<int>`) added to their name and I think `getDeclaringType()` / `getAMember()` will work correctly (finding all versions).

We actually have test cases for this already, e.g. the uses of `append` and `assign` with iterator parameters in `cpp\ql\test\library-tests\dataflow\taint-tests\string.cpp` around line 480.